File: app/src/main/java/com/noteapp/Model/NoteDatabase.java
package com.noteapp.Model;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import java.util.ArrayList;
import java.util.Date;

public class NoteDatabase extends SQLiteOpenHelper {

    // Allgemein gueltige public Konstanten
    private static final int TITLE_CC = 30;
    private static final int TEXT_CC = 200;
    private static final int DATE_CC = 20;
    // Allgemein gueltige private Konstanten
    private static final int DATABASE_VERSION = 1;
    private static final String DATABASE_NAME = "MyNoteDb";
    // Anzahl der Chars, muss-Angabe fuer VarChar2, zu String konvertiert
    private static final String TITLE_CCSTR = String.valueOf(TITLE_CC);
    private static final String TEXT_CCSTR = String.valueOf(TEXT_CC);
    private static final String DATE_CCSTR = String.valueOf(DATE_CC);

    // Tabellen- und deren Spaltennamen
    private static final String TABLE = "note", KEY_ID = "id", KEY_TITLE = "title", KEY_TEXT = "text", KEY_DATE = "date";

    // Konstruktor
    public NoteDatabase(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
    }

    private static void createTableNOTE(SQLiteDatabase db) {
        // 1. String titleTV | 2. String titleTV | 3. String / java.util date
        db.execSQL("CREATE TABLE " + TABLE + "(" +
                KEY_ID + " INTEGER PRIMARY KEY, " +
                KEY_TITLE + " VARCHAR(" + TITLE_CCSTR + "), " +
                KEY_TEXT + " VARCHAR(" + TEXT_CCSTR + "), " +
                KEY_DATE + " VARCHAR(" + DATE_CCSTR + ")" +
                ")");
    }

    private static void dropTableNOTE(SQLiteDatabase db) {
        db.execSQL("DROP TABLE IF EXISTS " + TABLE);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        createTableNOTE(db);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        dropTableNOTE(db);
        onCreate(db);
    }

    /*/////////////////
    CRUD Methods = Create, Read, Update, Delete
    Schreiben, Lesen, Updaten und Loeschen von Eintraegen
    *//////////////////
    public Long insertNote(Note noteEntity) {
        SQLiteDatabase db = getWritableDatabase();
        ContentValues values = new ContentValues();

        String birthStr = String.valueOf(noteEntity.getDate().getTime());
        values.put(KEY_TITLE, noteEntity.getTitle());
        values.put(KEY_TEXT, noteEntity.getText());
        values.put(KEY_DATE, birthStr);

        Long id = db.insertOrThrow(TABLE, null, values);

        Log.println(Log.ASSERT, "insertNote", "new  Id = " + id);

        noteEntity.setId(id.intValue());

        db.close();
        return id;
    }

    public boolean deleteNote(String idStr) {
        int oldNoteCount = this.getNoteCount();

        SQLiteDatabase db = getWritableDatabase();
        db.delete(TABLE, KEY_ID + " = ?", new String[]{idStr});
        db.close();

        int newNoteCount = this.getNoteCount();

        return (newNoteCount + 1) == oldNoteCount;
    }

    public ArrayList<Note> getAllNotes() {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM " + TABLE, null);

        ArrayList<Note> res = new ArrayList<>();

        if (cursor != null && cursor.moveToFirst()) {
            do {
                int id = cursor.getInt(0);
                String notetitle = cursor.getString(1);
                String noteText = cursor.getString(2);
                Date noteDate = new Date(Long.valueOf(cursor.getString(3)));

                res.add(new Note(id, notetitle, noteText, noteDate));

            } while (cursor.moveToNext());
            cursor.close();
        }
        db.close();

        return res;
    }

    private int getNoteCount() {
        SQLiteDatabase db = getReadableDatabase();
        Cursor cursor = db.rawQuery("SELECT * FROM " + TABLE, null);
        //"SELECT COUNT(*) FROM " + TABLE ????

        int count = cursor.getCount();
        db.close();
        cursor.close();

        return count;
    }
}

File: app/src/main/java/com/noteapp/Model/NoteDatabase.kt
package com.noteapp.Model

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import java.util.*

class NoteDatabase(context: Context) : SQLiteOpenHelper(context, NoteDatabase.DATABASE_NAME, null, NoteDatabase.DATABASE_VERSION) {

    companion object {

        // Allgemein gueltige public Konstanten
        private val TITLE_CC = 30
        private val TEXT_CC = 200
        private val DATE_CC = 20
        // Allgemein gueltige private Konstanten
        private val DATABASE_VERSION = 1
        private val DATABASE_NAME = "MyNoteDb"
        // Anzahl der Chars, muss-Angabe fuer VarChar2, zu String konvertiert
        private val TITLE_CCSTR = TITLE_CC.toString()
        private val TEXT_CCSTR = TEXT_CC.toString()
        private val DATE_CCSTR = DATE_CC.toString()

        // Tabellen- und deren Spaltennamen
        private val TABLE = "note"
        private val KEY_ID = "id"
        private val KEY_TITLE = "title"
        private val KEY_TEXT = "text"
        private val KEY_DATE = "date"

        private fun createTableNOTE(db: SQLiteDatabase) {
            // 1. String titleTV | 2. String titleTV | 3. String / java.util date
            db.execSQL("CREATE TABLE " + TABLE + "(" +
                    KEY_ID + " INTEGER PRIMARY KEY, " +
                    KEY_TITLE + " VARCHAR(" + TITLE_CCSTR + "), " +
                    KEY_TEXT + " VARCHAR(" + TEXT_CCSTR + "), " +
                    KEY_DATE + " VARCHAR(" + DATE_CCSTR + ")" +
                    ")")
        }

        private fun dropTableNOTE(db: SQLiteDatabase) {
            db.execSQL("DROP TABLE IF EXISTS " + TABLE)
        }
    }

    override fun onCreate(db: SQLiteDatabase) {
        createTableNOTE(db)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        dropTableNOTE(db)
        onCreate(db)
    }

    /*/////////////////
    CRUD Methods = Create, Read, Update, Delete
    Schreiben, Lesen, Updaten und Loeschen von Eintraegen
    *//////////////////
    fun insertNote(noteEntity: Note): Long? {
        val db = writableDatabase
        val values = ContentValues()

        val birthStr = noteEntity.date.time.toString()
        values.put(KEY_TITLE, noteEntity.title)
        values.put(KEY_TEXT, noteEntity.text)
        values.put(KEY_DATE, birthStr)

        val id = db.insertOrThrow(TABLE, null, values)

        Log.println(Log.ASSERT, "insertNote", "new  Id = " + id)

        noteEntity.id = id.toInt()

        db.close()
        return id
    }

    fun deleteNote(idStr: String): Boolean {
        val oldNoteCount = this.noteCount

        val db = writableDatabase
        db.delete(TABLE, KEY_ID + " = ?", arrayOf(idStr))
        db.close()

        val newNoteCount = this.noteCount

        return newNoteCount + 1 == oldNoteCount
    }

    fun updateNote(noteEntity: Note): Long? {
        val db = writableDatabase
        val values = ContentValues()

        val birthStr = noteEntity.date.time.toString()
        values.put(KEY_TITLE, noteEntity.title)
        values.put(KEY_TEXT, noteEntity.text)
        values.put(KEY_DATE, birthStr)

        val idStr = noteEntity.id.toString()
        val res = db.update(TABLE, values, KEY_ID + "= ?", arrayOf(idStr))

        Log.println(Log.ASSERT, "setUpdatingId", "setUpdatingId affected Rows: " + res)

        db.close()

        return res.toLong()
    }

    val allNotes: ArrayList<Note>
        get() {
            val db = readableDatabase
            val cursor = db.rawQuery("SELECT * FROM " + TABLE, null)

            val res = ArrayList<Note>()

            if (cursor != null && cursor.moveToFirst()) {
                do {
                    val id = cursor.getInt(0)
                    val notetitle = cursor.getString(1)
                    val noteText = cursor.getString(2)
                    val noteDate = Date(java.lang.Long.valueOf(cursor.getString(3))!!)

                    res.add(Note(id, notetitle, noteText, noteDate))

                } while (cursor.moveToNext())
                cursor.close()
            }
            db.close()

            return res
        }

    private //"SELECT COUNT(*) FROM " + TABLE ????
    val noteCount: Int
        get() {
            val db = readableDatabase
            val cursor = db.rawQuery("SELECT * FROM " + TABLE, null)

            val count = cursor.count
            db.close()
            cursor.close()

            return count
        }
}